stdlib/InteractiveUtils/src/macros.jl

```diff
@@ -37,20 +57,85 @@ function recursive_dotcalls!(ex, args, i=1)
     return ex, i
 end
 
+function extract_farg(arg)
+    !isexpr(arg, :(::), 1) && return esc(arg)
+    fT = esc(arg.args[1])
+    :($construct_callable($fT))
+end
+
+function construct_callable(@nospecialize(func::Type))
+    # Support function singleton types such as `(::typeof(f))(args...)`
+    Base.issingletontype(func) && isdefined(func, :instance) && return func.instance
+    # Don't support type annotations otherwise, we don't want to give wrong answers
+    # for callables such as `(::Returns{Int})(args...)` where using `Returns{Int}`
+    # would give us code for the constructor, not for the callable object.
+    throw(ArgumentError("If a function type is explicitly provided, it must be a singleton whose only instance is the callable object"))
+end
+
+function separate_kwargs(exs)
+    args = []
+    kwargs = []
+    for ex in exs
+        if isexpr(ex, :kw)
+            push!(kwargs, ex)
+        elseif isexpr(ex, :parameters)
+            for kw in ex.args
+                push!(kwargs, kw)
+            end
+        else
+            push!(args, ex)
+        end
+    end
+    args, kwargs
+end
+
+function namedtuple_type(kwargs)
+    names = Expr(:tuple)
+    types = Expr(:curly, :Tuple)
+    splats = Expr[]
+    for ex in kwargs
+        if isexpr(ex, :kw, 2)
+            name, value = ex.args[1], ex.args[2]
+            push!(names.args, QuoteNode(name::Symbol))
+            push!(types.args, get_typeof(value))
+        elseif isexpr(ex, :..., 1)
+            push!(splats, Expr(:call, typeof_nt, esc(ex.args[1])))
+        else
+            isa(ex, Symbol) || return nothing
+            push!(names.args, QuoteNode(ex))
+            push!(types.args, get_typeof(ex))
+        end
+    end
+    nt = :(NamedTuple{$names, $types})
+    isempty(splats) && return nt
+    length(splats) == 1 && return :($namedtuple_type($nt, $(splats[1])))
+    :(foldl($namedtuple_type, Any[$(splats...)]; init = $nt))
+end
+
+typeof_nt(nt::NamedTuple) = typeof(nt)
+typeof_nt(nt::Base.Pairs) = typeof(values(nt))
+
+function namedtuple_type(@nospecialize(nt1::Type{<:NamedTuple}), @nospecialize(nt2::Type{<:NamedTuple}))
+    names = tuple(fieldnames(nt1)..., fieldnames(nt2)...)
+    types = Tuple{fieldtypes(nt1)..., fieldtypes(nt2)...}
+    NamedTuple{names, types}
+end
+
 function gen_call_with_extracted_types(__module__, fcn, ex0, kws=Expr[])
-    if Meta.isexpr(ex0, :ref)
+    if isexpr(ex0, :ref)
         ex0 = replace_ref_begin_end!(ex0)
     end
     # assignments get bypassed: @edit a = f(x) <=> @edit f(x)
     if isa(ex0, Expr) && ex0.head == :(=) && isa(ex0.args[1], Symbol) && isempty(kws)
         return gen_call_with_extracted_types(__module__, fcn, ex0.args[2])
     end
     if isa(ex0, Expr)
-        if ex0.head === :do && Meta.isexpr(get(ex0.args, 1, nothing), :call)
+        ex0, where_params = extract_where_parameters(ex0)
+        if ex0.head === :do && isexpr(get(ex0.args, 1, nothing), :call)
```

Comment:
Probably better add guard against `!(ex0::Expr)` possibility, or add type assertion in `extract_where_parameters`.